### ISSUE TITLE
response.Body.FlushAsync does nothing until 1024 bytes written to body

### ISSUE THREAD

**Issue by The-MAZZTer** [aspnet/KestrelHttpServer#1092]
I searched through the source code to try and explain this behavior. Closest thing I found is: https://github.com/aspnet/KestrelHttpServer/issues/1061 ,which is a change for a future release I don't have.

I don't think that is what I'm seeing though, as that describes a 64k buffer and it applies to the whole stream, and it keeps buffering after the first write. I've confirmed what I'm seeing happens until 1k, never happens afterwards, and adding response headers does not affect it.

Here a minimal ActionResult's ExecuteResultAsync, which should be all you need to reproduce:
```
		public async override Task ExecuteResultAsync(ActionContext context) {
			HttpResponse response = context.HttpContext.Response;
			response.Headers.Add("Content-Encoding", "none");
			response.ContentType = "text/plain";
			while (true) {
				if (context.HttpContext.RequestAborted.IsCancellationRequested) {
					return;
				}

				try {
					await response.WriteAsync("1234567890");
					await response.Body.FlushAsync();
				} catch (Exception) {
					try {
						response.Body.Dispose();
					} catch (Exception) { }
				}

				await Task.Delay(100);
			}
		}
```

I've only recently discovered you can use async functions for routes... this is what I hacked together not knowing that. Should still work. Works fine past 1kb of content.

Only workaround I've figured out is to spit out 1kb of garbage data before any real data which isn't ideal.

I am using the .NET Core 1.0 release, running on Windows 7 x64. My project is using Kestrel.

[Edit: Going to try .NET Core source stepping assuming that's a thing you can do and see what the code is doing that way.]

[Edit: I traced the code all the way down to Microsoft.AspNetCore.Server.Kestrel.Internal.Networking.Libuv.write. Both calls seem identical before and after 1024 bytes have been written; I have no clue what's different.]

**Comment by benaadams** [comment]
I think I have identified the issue and raised a separate one that encapsulates it https://github.com/aspnet/KestrelHttpServer/issues/1093

Btw you don't need to Dispose the Body stream as it is a null operation; but you also probably shouldn't continue to infinitely loop after you have tried to dispose the Body stream.

**Comment by benaadams** [comment]
Basically `FlushAsync` completes immediately if below buffer limit as does `WriteAsync` so you are writing to the stream in a very very tight loop and likely write 1024 bytes on it before it gets a chance to output it.

**Comment by davidfowl** [comment]
> Btw you don't need to Dispose the Body stream as it is a null operation;

You shouldn't dispose streams you didn't create.

**Comment by benaadams** [comment]
> > Btw you don't need to Dispose the Body stream as it is a null operation;

> You shouldn't dispose streams you didn't create.

Ah, yes might do bad thing if was SSL, GZip or buffer stream

**Comment by The-MAZZTer** [comment]
Alright, thanks. I think I got that from StackOverflow, and I probably used .NET 4.5 code and modified it to work on .NET Core. I assumed at some point RequestAborted would get set. I'll replace the Dispose with a return statement.

Glad to hear you were able to reproduce and have an idea to fix. Is there a workaround I can do in the meantime? Something I can tweak through Reflection? Or something quick and dirty in the source I can modify and recompile?

**Comment by khellang** [comment]
> I assumed at some point `RequestAborted` would get set. I'll replace the Dispose with a return statement.

Just some info on `RequestAborted`; https://github.com/aspnet/Mvc/issues/5239#issuecomment-244991339

**Comment by benaadams** [comment]
> Glad to hear you were able to reproduce and have an idea to fix. Is there a workaround I can do in the meantime?

Flush will flush the stream; the issue is on the server you regain the ability to write before the flush is complete.

Eventually it will block and pause the writes so I'm not sure it will cause a huge issue it in that you shouldn't end up with a large backlog of items to be written; and the flush will be actioned.

**Comment by The-MAZZTer** [comment]
That is not the issue I filed this bug report about.

Whether I call Flush or not doesn't change the behavior of the sample code I posted; the problem I am seeing can be reproduced without it.

Scenarios I tested:
- I write out less than 1024 bytes to the response body and return from ExecuteResultAsync immediately. The output is then immediately sent to the browser and I can see it.
- I write out more than 1024 bytes right away to the response and don't return from ExecuteResultAsync, but endlessly loop. The output shows up in the browser right away. Any future WriteAsync calls show up right away.
- I write out less than 1024 bytes right away to the response and don't return from ExecuteResultAsync, but endlessly loop. The output does not show up in the browser until the output accumulates to 1024 bytes. Chrome's Network Developer Tool claims 0 bytes received on the connection until that point.

Other things I tested:
- Adding additional headers to the response does not change the required 1024 bytes in the body. This suggests to me whatever is happening is happening on the HTTP protocol handling level (or higher) rather than the socket level.
- Including or not including the FlushAsync call has no effect on the output.
- Using Body.WriteAsync instead of response.WriteAsync is no different.
- Substituting an async controller for a ActionResult gives the same result:

```
		[Route("test"), HttpGet]
		public async Task Test() {
			HttpResponse response = this.HttpContext.Response;
			response.Headers.Add("Content-Encoding", "none");
			response.ContentType = "text/plain";
			while (true) {
				if (this.HttpContext.RequestAborted.IsCancellationRequested) {
					return;
				}

				try {
					await response.WriteAsync("1234567890");
					await response.Body.FlushAsync();
				} catch {
					return;
				}

				await Task.Delay(100);
			}
		}
```

As a side note I am navigating to my API endpoints using Google Chrome and seeing this behavior (latest Stable and Dev both work).
